File: Problem_1.js
// Implement Trie (Prefix Tree)(https://leetcode.com/problems/implement-trie-prefix-tree/)

// Time Complexity

// Space Complexity


// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No
// Your code here along with comments explaining your approach

var TrieNode = function() {
    this.children = new Array(26);
    this.isEnd = false;
}

var Trie = function() {
    this.root = new TrieNode();
};

/** 
 * @param {string} word
 * @return {void}
 */
Trie.prototype.insert = function(word) {
    let curr = this.root;
    for(let i=0; i<word.length; i++){
        let ch = word[i];
        let index = ch.charCodeAt(0) - 'a'.charCodeAt(0);
        if(!curr.children[index]){
            curr.children[index] = new TrieNode();      
        }
        curr = curr.children[index];
    }
    curr.isEnd = true;
};

/** 
 * @param {string} word
 * @return {boolean}
 */
Trie.prototype.search = function(word) {
    let curr = this.root;
    for(let i=0; i<word.length; i++){
        let ch = word[i];
        let index = ch.charCodeAt(0) - 'a'.charCodeAt(0);
        if(!curr.children[index]){
            return false;
        }
        curr = curr.children[index];
    }
    return curr.isEnd;
};

/** 
 * @param {string} prefix
 * @return {boolean}
 */
Trie.prototype.startsWith = function(prefix) {
    let curr = this.root;
    for(let i=0; i<prefix.length; i++){
        let ch = prefix[i];
        let index = ch.charCodeAt(0) - 'a'.charCodeAt(0);
        if(!curr.children[index]){
            return false;
        }
        curr = curr.children[index];
    }
    return true;
};

/** 
 * Your Trie object will be instantiated and called as such:
 * var obj = new Trie()
 * obj.insert(word)
 * var param_2 = obj.search(word)
 * var param_3 = obj.startsWith(prefix)
 */

File: Problem1.cs
public class Trie
    {
        // Time Complexity : Insert : O(l), Search(): O(l), StartsWith(): O(l) - where l is the length of the word
        // Space Complexity : Insert : O(n*l), Search(): O(n*l), StartsWith(): O(n*l) - where n is the number of words and l is the length of each word
        // Did this code successfully run on Leetcode : Yes
        // Any problem you faced while coding this : No
        TrieNode root;

        public class TrieNode
        {
            public bool isEnd;
            public TrieNode[] children;

            public TrieNode()
            {
                children = new TrieNode[26];
            }
        }
        public Trie()
        {
            root = new TrieNode();   
        }

        public void Insert(string word)
        {
            TrieNode curr = root;
            for(int i = 0; i < word.Length; i++)
            {
                char c = word[i];
                if (curr.children[c - 'a'] == null)
                {
                    curr.children[c - 'a'] = new TrieNode();
                }
                curr = curr.children[c - 'a'];
            }
            curr.isEnd = true;
        }

        public bool Search(string word)
        {
            TrieNode curr = root;
            for(int i = 0; i < word.Length; i++)
            {
                char c = word[i];
                if (curr.children[c - 'a'] == null)
                {
                    return false;
                }
                curr = curr.children[c - 'a'];
            }
            return curr.isEnd;
        }

        public bool StartsWith(string prefix)
        {
            TrieNode curr = root;
            for(int i = 0; i < prefix.Length; i++)
            {
                char c = prefix[i];
                if (curr.children[c - 'a'] == null)
                    return false;

                curr = curr.children[c - 'a'];
            }
            return true;
        }
    }
